analyses/copy_number_consensus_call/src/scripts/compare_variant_calling_updated.py

```diff
@@ -145,115 +108,183 @@
+                        overlap_chrom = [chr_list1,str(chrom_start),str(chrom_end),'NULL',str(cnv_list1).strip(','),list2_chr_str_end.strip(','),m[-1]]
+
+                    ## Add the formatted consensus CNV into a growing list of consensus CNVs
+                    consensus_list = consensus_list + [overlap_chrom]
+
+    return consensus_list
+
+
+## Function to save output files
+def save_to_file(output_file_content, output_path, sample_name):
+    ## Open up the file to write in it, the 'w' option overwrites the file if the file exists.
+    with open( output_path, 'w') as file:
+        ## Loop through the output file and print line by line
+        for k in output_file_content:
+
+            ## Add the sample name and file name to each line
+            single_name = output_path.split('/')[-1]
+            file.write('\t'.join(k) + '\t'+ sample_name + '\t' + single_name +'\n')
+    sys.stderr.write('$$$ Write to file ' + str(output_path) + ' was sucessful\n')
+
+
+
+################### SCRIPT BODY #################################
+
+
+
+## Define parser for the input file
+parser = argparse.ArgumentParser(description="""This script takes in 3 bed files, each from one of the 3 callers
+                                                 and find common CNVs between two files at a time.""")
+parser.add_argument('--manta', required=True,
+                    help='path to the manta file')
+parser.add_argument('--cnvkit', required=True,
+                    help='path to the cnvkit file')
+parser.add_argument('--freec', required=True,
+                    help='path to the freec file')
+parser.add_argument('--manta_cnvkit', required=True,
+                    help='path of the output consensus between manta and cnvkit')
+parser.add_argument('--manta_freec', required=True,
+                    help='path of the output consensus between manta and freec')
+parser.add_argument('--cnvkit_freec', required=True,
+                    help='path of the output consensus between cnvkit and freec')
+parser.add_argument('--sample', default='no_sample',
+                    help='sample name to use in the file')
+
+args = parser.parse_args()
+
+
+## Read in the input files as dictionaries
+manta_dict = read_input_file(args.manta)
+cnvkit_dict = read_input_file(args.cnvkit)
+freec_dict = read_input_file(args.freec)
+
+## Put the input dictionaries into a bigger dictionary
+dict_of_input_content = {'manta':manta_dict, 'cnvkit':cnvkit_dict, 'freec':freec_dict}
+
+input_file_names = list(dict_of_input_content.keys())
+
+## Put the output file paths into their own lists that is to be iterated over
+## This order is important
+output_list = [args.manta_cnvkit, args.manta_freec, args.cnvkit_freec]
+
+## Make a list of iteraby index for the MAIN for-loop below
+## This give a list as followed: [0,1,2]
+list_index = list(range(0,len(dict_of_input_content)))
```

Comment:
Rather than doing these lists and indexes, we could simplify this to something like this to keep it simpler
```
caller_pairs = [('manta', 'cnvkit'), ('manta', 'freec'), ('cnvkit', 'freec')]
```

Then you can use a single loop below, something like:
```
for caller1, caller2 in caller_pairs:
    list1 = dict_of_input_content[caller1]
    list2 = dict_of_input_content[caller2]
    ....
```

Reply:
Thank you for the suggestion @jashapiro. The reason why I did it my way was because I was trying to avoid hardcoding as much as possible. My intention was that when someone want to change the code, they can go in and be able to change the code more easily. 

I guess what I am struggling with is when to hardcode this information and when not to. However, I do agree that your method is simpler and definitely more clear. I will change the script to reflect that.

Reply:
In this case it might be even better to try itertools.combinations. Does `itertools.combinations(['manta', 'cnvkit', 'freec'], 2)` do what you want in terms of getting caller pairs that you can iterate though?

I also wonder if you could use `itertools.combinations` over dict_of_input_content's keys to get something that's both clear and more general. I don't quite understand what jval and kval are so I can't provide feedback that's highly specific to this code.

Reply:
Link for convenience to itertools.combinations: https://docs.python.org/3/library/itertools.html#itertools.combinations

Reply:
The only potential issue I see with this is if the order is not as expected: some of the surrounding code (in other scripts/snakefile) does depend on filenames being as expected (with the expected content!).

Reply:
> Thank you for the suggestion @jashapiro. The reason why I did it my way was because I was trying to avoid hardcoding as much as possible. My intention was that when someone want to change the code, they can go in and be able to change the code more easily.

I think this is a good goal, but given that there is already some hard-coding in here (notably the arguments), it is going to be hard to get rid of it all. 

If removing hard coding were the full goal, I would probably simplify this script to take in two files and output the merged file, then move the combinations logic to snakemake. That would lose some efficiency because you would have to read in data more than once, but it is easier to extend. Having already moved the main merge logic to a function, this would be easy to do in the future if desired.